drivers/ethernet/eth_e1000.c

```diff
@@ -0,0 +1,340 @@
+{
+	return  ETHERNET_LINK_10BASE_T | ETHERNET_LINK_100BASE_T | \
+		ETHERNET_LINK_1000BASE_T;
+}
+
+static size_t e1000_linearize(struct net_pkt *pkt, void *buf, size_t bufsize)
+{
+	size_t len = net_pkt_ll_reserve(pkt) + pkt->frags->len;
+	struct net_buf *nb;
+
+	/* First fragment contains link layer (Ethernet) header */
+	memcpy(buf, net_pkt_ll(pkt), len);
+
+	for (nb = pkt->frags->frags; nb; nb = nb->frags) {
+		memcpy(buf + len, nb->data, nb->len);
+		len += nb->len;
+	}
+
+	return len;
+}
+
+static int e1000_tx(struct e1000_dev *dev, void *data, size_t data_len)
+{
+	dev->tx.addr = POINTER_TO_INT(data);
+	dev->tx.len = data_len;
+	dev->tx.cmd = TDESC_EOP | TDESC_RS;
+
+	W(dev, TDT, 1);
+
+	while(!(dev->tx.sta)) {
+		k_yield();
+	}
+
+	D("tx.sta: 0x%02hx", dev->tx.sta);
+
+	return (dev->tx.sta & TDESC_STA_DD) ? 0 : -EIO;
+}
+
+static int e1000_send(struct net_if *iface, struct net_pkt *pkt)
+{
+	struct e1000_dev *dev = net_if_get_device(iface)->driver_data;
+
+	size_t len = e1000_linearize(pkt, dev->txb, sizeof(dev->txb));
+
+	net_pkt_unref(pkt);
+
+	return e1000_tx(dev, dev->txb, len);
+}
+
+static struct net_pkt *e1000_rx(struct e1000_dev *dev)
+{
+	struct net_pkt *pkt = NULL;
+
+	D("rx.sta: 0x%02hx", dev->rx.sta);
+
+	if (!(dev->rx.sta & RDESC_STA_DD)) {
+		E("RX descriptor not ready");
+		goto out;
+	}
+
+	pkt = net_pkt_get_reserve_rx(0, K_NO_WAIT);
+	if (!pkt) {
+		E("Out of RX buffers");
+		goto out;
+	}
+
+	union {
+		u64_t qw;
+		u32_t dw[2];
+	} addr = { .qw = dev->rx.addr };
+
+	if (!net_pkt_append_all(pkt, dev->rx.len - 4,
+			INT_TO_POINTER(addr.dw[0]), K_NO_WAIT)) {
+		E("Out of memory for received frame");
+		net_pkt_unref(pkt);
+		pkt = NULL;
+	}
+out:
+	return pkt;
+}
+
+static void e1000_isr(struct device *device)
+{
+	struct e1000_dev *dev = device->driver_data;
+	u32_t icr = R(dev, ICR); /* Cleared upon read */
+
+	icr &= ~(ICR_TXDW | ICR_TXQE);
+
+	if (icr & ICR_RXO) {
+		struct net_pkt *pkt = e1000_rx(dev);
+
+		icr &= ~ICR_RXO;
+
+		if (pkt) {
+			net_recv_data(dev->iface, pkt);
+		}
+	}
+
+	if (icr) {
+		E("Unhandled interrupt, ICR: 0x%x", icr);
+	}
+}
+
+int e1000_probe(struct device *device)
+{
+	struct e1000_dev *dev = device->driver_data;
+	bool found = false;
+
+	pci_bus_scan_init();
+
+	if (pci_bus_scan(&dev->pci)) {
+
+		pci_enable_regs(&dev->pci);
+
+		pci_enable_bus_master(&dev->pci);
+
+		pci_show(&dev->pci);
+
+		found = true;
+	}
+
+	return found;
+}
+
+#define PCI_VENDOR_ID_INTEL	0x8086
+#define PCI_DEVICE_ID_I82540EM	0x100e
+
+static struct e1000_dev e1000_dev = {
+	.pci.vendor_id = PCI_VENDOR_ID_INTEL,
+	.pci.device_id = PCI_DEVICE_ID_I82540EM,
+};
+
+static const struct ethernet_api e1000_api = {
+	.iface_api.init		= e1000_init,
+	.iface_api.send		= e1000_send,
+	.get_capabilities	= e1000_caps,
+};
+
+NET_DEVICE_INIT(eth_e1000,
+		"ETH_0",
+		e1000_probe,
+		&e1000_dev,
+		NULL,
+		CONFIG_ETH_INIT_PRIORITY,
+		&e1000_api,
+		ETHERNET_L2,
+		NET_L2_GET_CTX_TYPE(ETHERNET_L2),
+		E1000_MTU);
```

Comment:
Always place the NET_DEVICE_INIT macro at the EOF (and thus remove the declaration od e1000_init in the beginning of the file.)

Reply:
Since eth_e1000 is instantiated in the NET_DEVICE_INIT() macro, the IRQ_CONNECT() should be done below it. This would require one more function below NET_DEVICE_INIT() and a declaration for it. So, at the end NET_DEVICE_INIT() won't be at the end of file. I suggest to leave it as is.

Reply:
Change it, there is a good reason I am asking this: we are going to replace this macro style stuff by built-time generated code at some point, and all drivers doing irq connect as you do (a minor amount of drivers do that yes) are going to give use more work.